apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableMutateRowsRequestManager.cs

```diff
@@ -0,0 +1,254 @@
+            {
+                codes.Add(GetGrpcCode(status));
+            }
+            return codes;
+        }
+
+        /// <summary>
+        /// Adds the content of the MutateRowsResponse message to the <see cref="Results"/>
+        /// </summary>
+        /// <param name="response"></param>
+        public void SetStatus(MutateRowsResponse response)
+        {
+            foreach (var entry in response.Entries)
+            {
+                int index = (int)entry.Index;
+
+                if (index >= _mapToOriginalIndex.Length || index < 0)
+                {
+                    _messageIsInvalid = true;
+                    break;
+                }
+                // Set the result
+                Results[_mapToOriginalIndex[index]] = entry.Status;
+            }
+        }
+
+        /// <summary>
+        /// This method is called when all calls to <see cref="SetStatus(MutateRowsResponse)"/> are complete.
+        /// </summary>
+        /// <returns>
+        /// ProcessingStatus of the accumulated responses - success, invalid, retryable, non-retryable.
+        /// </returns>
+        public ProcessingStatus OnOk()
+        {
+            // Sanity check to make sure that every mutation received a response.
+            if (!_messageIsInvalid)
+            {
+                for (int i = 0; i < Results.Length; i++)
+                {
+                    if (Results[i] == null)
+                    {
+                        _messageIsInvalid = true;
+                        break;
+                    }
+                }
+            }
+
+            // There was a problem in the data found in SetStatus(), so fail the RPC.
+            if (_messageIsInvalid)
+            {
+                return ProcessingStatus.INVALID;
+            }
+
+            // This list hold indices of entries to be included in retry request
+            List<int> toRetry = new List<int>();
+            ProcessingStatus processingStatus = ProcessingStatus.SUCCESS;
+
+            // Check the current state to determine the state of the results.
+            // There are three states: OK, Fail, or Partial Retry.
+            for (int i = 0; i < Results.Length; i++)
+            {
+                Rpc.Status status = Results[i];
+                if (status.Code == GetGrpcCode(Grpc.Core.StatusCode.OK))
+                {
+                    continue;
+                }
+                if (_rertriableCodes.Contains(status.Code))
+                {
+                    // An individual mutation failed with a retryable code, usually DEADLINE_EXCEEDED.
+                    toRetry.Add(i);
+                    if (processingStatus == ProcessingStatus.SUCCESS)
+                    {
+                        processingStatus = ProcessingStatus.RETRYABLE;
+                    }
+                }
+                else
+                {
+                    // Don't retry if even a single response is not retryable.
```

Comment:
'_modifiedRequest' is created on constructor and initially is equal to '_originalRequest'.
If we will stay with not retrying in case of a single NotRetryable status, than I suggest to nullify or clear `_modifiedRequest` and clear `toRetry`. Otherwise `GetRetryRequest()` at this point will still return full `_modifiedRequest == _originalRequest`

Reply:
Or perhaps `RetryRequest` (like I said, that should be a property and `_modifiedRequest` should be removed) should be null always until the processing status is `RETRYABLE`. Then `RetryRequest`, as the name would suggest, only holds the request that should be made on a subsequent retry request. How does that sound?

Reply:
Based on my conversation with Solomon and after doing some tests I think we should probably stick to not retrying even if one response is not retryable. Details in the email